developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,201 @@
+property of some resource--it is said to _depend_ on that resource.
+
+If a value of type `Output<T>` depends on a resource `R`, the result of any computation
+that depends on its concrete value also depends on `R`.
+
+### Unknowns 
+
+An `Output<T>` may be unknown if it depends on the result of a resource operation that
+will not be run because it is part of a `pulumi preview`. Previews typically produce for
+properties with values that cannot be determined until the resource is actually created or
+updated.
+
+If a value of type `Output<T>` is unknown, any computation that depends on its concrete
+value must not run, and may produce an unknown `Output<T>`.
+
+### Secrets
+
+An `Output<T>` may be makred as secret if its concrete value contains sensitive
+information.
+
+If a value of type `Output<T>` is secret, the result of any computation that depends on
+its concrete value must also be secret.
+
+## `Input<T>`
+
+The partner of `Output<T>` is `Input<T>`, which is defined as `Union<T, Output<T>>`. In
+simpler terms, a location of type `Input<T>` may accept either a plain old `T` value or an
+`Output<T>` value.
+
+## `inputShape(T)`
+
+Although `Input<T>` gives us the ability to deal in both `T` and `Output<T>` values, it is
+often the case that we want to construct _composite_ values out of multiple `Input<T>`s.
+For example, consider `Input<Array<string>>`: a value of this type accepts either a
+`Array<string>` or an `Output<Array<string>>`, but does not accept a value of type
+`Array<Output<string>>`. In order to accept all three of these types, we need the type
+`Union<Input<Array<string>>>, Array<Input<string>>>`. The `inputShape` type constructor
+defines an algorithm for producing these sorts of values.
+
+```rust
+fn inputShape(T) {
+	match T {
+		_ => Input<T>,
+		Array<T> => Union<Input<Array<T>>, Array<inputShape(T)>>,
+		Map<T> => Union<Input<Map<T>>, Map<inputShape(T)>>,
+		Union<...T> => Union<map(...T, t => inputShape(t))>,
+		Promise<T> => Input<T>,
+		Output<T> => Input<T>,
+		Object<...P> => Object<map(...P, (name, t) => (name, inputShape(t)))>
+	}
+}
+```
+
+Resource input properties often use input-shaped types.
+
+## `Output<T>` Combinators
+
+The rules described for working with `Output<T>` metadata--[dependencies], [unknowns], and
+[secrets]--require special bookkeeping on the part of the consumer. There are three
+primitive combinators that aid in this bookkeeping: [`apply`](#applyt-u), [`all`](#allt),
+and [`unwrap`](#unwrapt).
+
+### `apply<T, U>`
+
+The `apply` API allows its caller to access the concrete value of an `Output<T>` within
+the context of a caller-supplied callback.
+
+`apply` trivially obeys the `Output<T>` rules for [dependencies], [unknowns], and
+[secrets]:
+
+- the dependencies of the `Output<T>` argument are propagated to the result
+- if the `Output<T>` argument is unknown, the callback is not run and the result is unknown
+- if the `Output<T>` argument is secret, the result is secret.
```

Comment:
I'm not sure I would put a period here. I think it's a grey area.

Reply:
Removed.